### ISSUE TITLE
Drush 9 sql:cli not working with xdebug enabled

### ISSUE THREAD

**Issue by mike-potter** [docksal/docksal#795]
**Description**

I have Drush 9 being brought into my project locally via composer.  When I go into the cli container using ``fin bash`` and then try using ``drush sql:cli`` it just hangs.  If I do ``drush sql:cli -vvv`` I get the following (it hands on the proc_open):

```
docker@cli:/var/www$ drush sqlc -vvv
 [preflight] Config paths: /var/www/vendor/drush/drush/drush.yml
 [preflight] Alias paths: /var/www/build/html/drush/sites,/var/www/drush/sites
 [preflight] Commandfile search paths: /var/www/vendor/drush/drush/src
 [bootstrap] Starting bootstrap to max [0.2 sec, 7.54 MB]
 [bootstrap] Drush bootstrap phase: bootstrapDrupalRoot() [0.22 sec, 7.73 MB]
 [bootstrap] Change working directory to /var/www/build/html [0.22 sec, 7.73 MB]
 [bootstrap] Initialized Drupal 8.6.1 root directory at /var/www/build/html [0.22 sec, 7.73 MB]
 [bootstrap] Drush bootstrap phase: bootstrapDrupalSite() [0.23 sec, 8.25 MB]
 [bootstrap] Initialized Drupal site default at sites/default [0.23 sec, 8.25 MB]
 [bootstrap] Drush bootstrap phase: bootstrapDrupalConfiguration() [0.23 sec, 8.25 MB]
 [debug] Add service modifier [0.24 sec, 8.41 MB]
Calling proc_open(mysql --defaults-file=/tmp/drush_KnzXEd --database=drupal_octane --host=db);
```

```
docker@cli:/var/www$ which drush
/usr/local/bin/drush
```

```
docker@cli:/var/www$ ls -al /usr/local/bin | grep drush
-rwxr-xr-x 1 root staff   310920 Jul 31 01:39 drush
-rwxr-xr-x 1 root staff  4590001 Jul 31 01:39 drush8
```

So I see that Docksal has 2 versions of drush in /usr/local/bin and would seem to indicate that "drush" was different from "drush8".  If I am outside my docroot and do "drush version" it shows "8.1.16" and if I do "drush8 version" is shows the same "8.1.16".

Within my docroot, "drush version" shows "9.4.0" which is the correct version from composer.  And if I directly run ``./vendor/drush/drush/drush sql:cli`` then it works correctly!

When I searched for drush help on this, they talk about issues having a global version of drush that is a different version from the vendor directory.  What I'm wondering is what Docksal is specifically doing in the cli container with the global drush version and if anybody else is running into this.

The issue for global drush version was here: https://github.com/drush-ops/drush/issues/1712 but seemed to be from the old drush 8 and claims to have been fixed.

So, I'm not sure if this is a drush issue, or an issue with how drush is being installed into the cli container.

Problem/feature description...

**Output of `fin config`:**

<details>
  <summary>fin config output</summary>

```
% fin config
---------------------
COMPOSE_PROJECT_NAME_SAFE: temp
COMPOSE_FILE:
/Users/michaelpotter/.docksal/stacks/overrides-osxfs.yml
/Users/michaelpotter/.docksal/stacks/volumes-bind.yml
/Users/michaelpotter/.docksal/stacks/stack-default.yml
/Users/michaelpotter/Projects/temp/.docksal/docksal.yml
/Users/michaelpotter/.docksal/stacks/overrides-ide.yml
ENV_FILE:
/Users/michaelpotter/Projects/temp/.docksal/docksal.env

PROJECT_ROOT: /Users/michaelpotter/Projects/temp
DOCROOT: build/html
VIRTUAL_HOST: temp.docksal
VIRTUAL_HOST_ALIASES: *.temp.docksal
IP: 192.168.64.100
MYSQL: 192.168.64.100:32774

Docker Compose configuration
---------------------
services:
  cli:
    dns:
    - 192.168.64.100
    - 8.8.8.8
    environment:
      DOCKER_ENV: local
      DOCROOT: build/html
      HOST_GID: '20'
      HOST_UID: '501'
      PHP_IDE_CONFIG: serverName=temp.docksal
      SECRET_ACAPI_EMAIL: null
      SECRET_ACAPI_KEY: null
      SECRET_BLACKFIRE_CLIENT_ID: null
      SECRET_BLACKFIRE_CLIENT_TOKEN: null
      SECRET_PLATFORMSH_CLI_TOKEN: null
      SECRET_SSH_PRIVATE_KEY: null
      SECRET_TERMINUS_TOKEN: null
      VIRTUAL_HOST: temp.docksal
      XDEBUG_CONFIG: remote_connect_back=0 remote_host=192.168.64.1
      XDEBUG_ENABLED: '1'
    hostname: cli
    image: docksal/cli:2.4-php7.1-ide
    labels:
      io.docksal.cert-name: none
      io.docksal.virtual-host: ide.temp.docksal,ide.temp.docksal.*
      io.docksal.virtual-port: '3000'
    volumes:
    - docksal_ssh_agent:/.ssh-agent:ro
    - cli_home:/home/docker:rw
    - /Users/michaelpotter/Projects/temp:/var/www:rw,cached
  db:
    dns:
    - 192.168.64.100
    - 8.8.8.8
    environment:
      MYSQL_ALLOW_EMPTY_PASSWORD: null
      MYSQL_DATABASE: drupal_octane
      MYSQL_INITDB_SKIP_TZINFO: null
      MYSQL_ONETIME_PASSWORD: null
      MYSQL_PASSWORD: admin
      MYSQL_RANDOM_ROOT_PASSWORD: null
      MYSQL_ROOT_PASSWORD: root
      MYSQL_USER: admin
    hostname: db
    image: docksal/db:1.2-mysql-5.6
    ports:
    - 0:3306/tcp
    volumes:
    - db_data:/var/lib/mysql:rw
    - project_root:/var/www:ro,nocopy
  web:
    depends_on:
      cli:
        condition: service_started
    dns:
    - 192.168.64.100
    - 8.8.8.8
    environment:
      APACHE_BASIC_AUTH_PASS: null
      APACHE_BASIC_AUTH_USER: null
      APACHE_DOCUMENTROOT: /var/www/build/html
      DOCKER_ENV: local
    hostname: web
    image: docksal/web:2.1-apache2.4
    labels:
      io.docksal.cert-name: none
      io.docksal.project-root: /Users/michaelpotter/Projects/temp
      io.docksal.virtual-host: temp.docksal,*.temp.docksal,temp.docksal.*
    volumes:
    - project_root:/var/www:ro,nocopy
version: '2.1'
volumes:
  cli_home: {}
  db_data: {}
  docksal_ssh_agent:
    external: true
    name: docksal_ssh_agent
  project_root:
    driver: local
    driver_opts:
      device: /Users/michaelpotter/Projects/temp
      o: bind
      type: none

---------------------
```

</details>

<br>

**Output of `fin sysinfo`:**

<details>
  <summary>fin sysinfo output</summary>

  ```
 % fin sysinfo
███  OS
Darwin Mac OS X 10.13.6
Darwin Michaels-MacBook-Pro.local 17.7.0 Darwin Kernel Version 17.7.0: Fri Jul  6 19:54:51 PDT 2018; root:xnu-4570.71.3~2/RELEASE_X86_64 x86_64

███  ENVIRONMENT
MODE : Docker for Mac

███  FIN
fin version: 1.69.1

███  DOCKER COMPOSE
EXPECTED VERSION: 1.21.1
docker-compose version 1.22.0, build f46880f
docker-py version: 3.4.1
CPython version: 3.6.4
OpenSSL version: OpenSSL 1.0.2o  27 Mar 2018

███  DOCKER
EXPECTED VERSION: 18.03.1-ce

Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:21:31 2018
 OS/Arch:           darwin/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:29:02 2018
  OS/Arch:          linux/amd64
  Experimental:     true

███  DOCKSAL: PROJECTS
project             STATUS              virtual host                                 project root
temp                Up 10 minutes       temp.docksal,*.temp.docksal,temp.docksal.*   /Users/michaelpotter/Projects/temp

███  DOCKSAL: VIRTUAL HOSTS
http://*.temp.docksal
http://ide.temp.docksal.*
http://ide.temp.docksal
http://temp.docksal.*
http://temp.docksal

███  DOCKER: RUNNING CONTAINERS
CONTAINER ID        IMAGE                        COMMAND                  CREATED             STATUS                    PORTS                                                    NAMES
de549cd12de0        docksal/web:2.1-apache2.4    "httpd-foreground"       10 minutes ago      Up 10 minutes             80/tcp, 443/tcp                                          temp_web_1
164a121d3301        docksal/cli:2.4-php7.1-ide   "/opt/startup.sh sup…"   10 minutes ago      Up 10 minutes (healthy)   22/tcp, 3000/tcp, 9000/tcp                               temp_cli_1
0547a8836791        docksal/ssh-agent:1.0        "/run.sh ssh-agent"      16 minutes ago      Up 16 minutes                                                                      docksal-ssh-agent
54ec05f38ebc        docksal/dns:1.0              "/opt/entrypoint.sh …"   16 minutes ago      Up 16 minutes             192.168.64.100:53->53/udp                                docksal-dns
e4dd47832b7a        docksal/vhost-proxy:1.3      "docker-entrypoint.s…"   16 minutes ago      Up 16 minutes             192.168.64.100:80->80/tcp, 192.168.64.100:443->443/tcp   docksal-vhost-proxy
5f14372cae8b        docksal/db:1.2-mysql-5.6     "/entrypoint.sh mysq…"   26 hours ago        Up 10 minutes             0.0.0.0:32774->3306/tcp                                  temp_db_1

███  DOCKER: NETWORKS
NETWORK ID          NAME                DRIVER              SCOPE
032ae12f78b0        _default            bridge              local
9992ee5750f4        bridge              bridge              local
046e164056ec        host                host                local
ebc10e967fd5        none                null                local
8a9eb4c8be21        temp_default        bridge              local

███  HDD Usage
Filesystem                                                                          Size   Used  Avail Capacity     iused               ifree %iused  Mounted on
/dev/disk1s1                                                                       932Gi  354Gi  572Gi    39%     5857231 9223372036848918576    0%   /
devfs                                                                              245Ki  245Ki    0Bi   100%         848                   0  100%   /dev
/dev/disk1s4                                                                       932Gi  5.0Gi  572Gi     1%           5 9223372036854775802    0%   /private/var/vm
map -hosts                                                                           0Bi    0Bi    0Bi   100%           0                   0  100%   /net
map auto_home                                                                        0Bi    0Bi    0Bi   100%           0                   0  100%   /home
  ```

</details>

**Comment by lmakarov** [comment]
@mike-potter `/usr/local/bin/drush` is Drush Launcher. It will pass controls to the project level drush (if installed), otherwise, it will use the binary set in `DRUSH_LAUNCHER_FALLBACK`.
`DRUSH_LAUNCHER_FALLBACK=/usr/local/bin/drush8` in `docksal/cli`.

Try `unset DRUSH_LAUNCHER_FALLBACK` and then run `drush sql:cli` again. If this still does not work, then the drush8 fallback is not the one to blame. There then must be an issue in the way Drush Launcher works with your project level Drush version.

**Comment by mike-potter** [comment]
Unsetting the ``DRUSH_LAUNCHER_FALLBACK`` didn't help.  Is docksal using the latest Drush launcher?

**Comment by lpeabody** [comment]
Everytime I see Drush hang like this, I always have to double check Xdebug in PHPStorm and make sure I've selected more than 1 in the max simultaneous connections field... that isn't it by chance is it? I'm like 99% sure that's not the problem here, but in the 1% chance it is... does it hang on any other Drush commands?

**Comment by mike-potter** [comment]
I originally checked that because I've definitely had times in the past where PHPStorm was waiting in debug mode.  But ``drush status`` works fine, and it also still hangs even when PHPStorm isn't running.

However, based on that idea I set ``XDEBUG_ENABLED=0`` in docksal.env and then reset the cli container and that fixed it!

So somehow xdebug is interfering with drush even when PHPStorm isn't running.

I've been using Docksal for a while with Drush 8 and never had this problem.  I noticed that command line debugging in Drush 9 is also not working and I have to use ``fin exec drush`` rather than ``fin drush`` when I do want to debug.  So seems like this might be a Drush 9 issue, but interested if others have run into this.

**Comment by lpeabody** [comment]
It might be possible that you update `XDEBUG_ENABLED=0` and forgot to run `fin up` :) I do that all the time... But yeah, Drush 9 has issues with command line debugging. I've been using `fin exec -T /var/www/vendor/bin/drush ...` and that seems to work. Slightly inconvenient to do the full path, but at least BLT-issued Drush commands are done using the full path.

At any rate, glad it's fixed for ya!

**Comment by lpeabody** [comment]
I just ran the same test. Turned `XDEBUG_ENABLED=1`, `fin bash`, `drush sql:cli` and it hangs for me also, without PHPStorm listening for XDebug. Interesting! At least we can duplicate it. This is on version php7.1-2.2 of service-cli.

**Comment by lmakarov** [comment]
> I've been using fin exec -T /var/www/vendor/bin/drush ... and that seems to work. 

@lpeabody does this also work with `XDEBUG_ENABLED=1`, but PHPStorm not listening to a debug session? If so, then drush itself is fine with xdebug. It's the launcher that causes things to hang.

A little off-topic: 
> This is on version php7.1-2.2 of service-cli.

Is there a particular reason you are sticking with an older version of `docksal/cli` and/or Docksal?

**Comment by lpeabody** [comment]
> @lpeabody does this also work with XDEBUG_ENABLED=1, but PHPStorm not listening to a debug session? If so, then drush itself is fine with xdebug. It's the launcher that causes things to hang.

It DOES NOT hang when specifying the full path to the Drush executable. Looks like Drush launcher is causing issues.

> Is there a particular reason you are sticking with an older version of docksal/cli and/or Docksal?

Laziness. :(

**Comment by achekulaev** [comment]
I wonder if this ticket can progress somewhere really

**Comment by sean-e-dietrich** [comment]
I have confirmed what @lpeabody mentioned about if you reference the path to the Drush 9 executable it works, but using the Drush launcher is where there are hangups.

Going to open a ticket on the launcher project.

**Comment by weitzman** [comment]
FYI this is fixed in Drush 9.6, which is in release candidate stage.

**Comment by sean-e-dietrich** [comment]
@lmakarov @achekulaev I'm thinking we can close this as it's been handled and will be updated in the next version of drush 9.
